### PR TITLE
[Work-in-Progress] Endpoints to allow downloading reports as CSV (#288, #289, #290)

### PR DESCRIPTION
**Pull request by Deep-Chill**
Added endpoints to allow downloading reports as CSV.

Some reports need their data to be slightly transformed, eg. balances-due view requires accounting for different units, the coupon_uses view requires accounting for nested values.

Notes:
- Works with query parameters.
- Tried the content-type approach with the request header sending Accept:text/csv but that didn't work with DRF, I think it could work with a custom CSV renderer.

### CODE REVIEW

**Comment by Deep-Chill** [comment]
Added #289 here because it uses the Mixin created in #288 and #290 was a minor change. Not sure if the download views/URLs belong with the API views/URLs, let me know if you want them moved to 'urls/views'.

File: saas/mixins.py [hunk 1]
```diff
@@ -1249,6 +1250,42 @@ def decorate_queryset(self, queryset):
         return decorated_queryset
 
 
+class CSVWriterMixin:
```
Comment (smirolo) [inline saas/mixins.py:1253]: Why define a `CSVWriteMixin` instead of using `CSVDownloadView`?

File: saas/api/metrics.py [hunk 2]
```diff
@@ -837,3 +837,230 @@ def paginate_queryset(self, queryset):
         page = super(BalancesDueAPIView, self).paginate_queryset(
             queryset)
         return self.decorate_queryset(page if page else queryset)
+
+
+class BalancesDownloadView(CSVWriterMixin, MetricsMixin, APIView):
+
+    def retrieve_metrics(self, date_periods):
```
Comment (smirolo) [inline saas/api/metrics.py:844]: It seems the `retrieve_metrics` shouldn't be redefined (as a copy/paste of the API version) but rather a `BalancesMetricsMixin` should be extracted out of `BalancesAPIView` that can be used for both, the API and Download.

**Review by smirolo — CHANGES_REQUESTED** [review]
Needs to follow current code pattern for API/Download implementation.

File: saas/api/metrics.py [hunk 3]
```diff
@@ -56,23 +56,178 @@ class MetricsMixin(DateRangeContextMixin, ProviderMixin):
...
+                'values': values,
+            })
+
+        extra_values = churn_subscribers_by_period(date_periods=date_periods)
+        extra = [{
+            'slug': 'churn',
+            'values': extra_values
+        }]
+        return table, extra
+    def retrieve_metrics(self, request):
```
Comment (smirolo) [inline saas/api/metrics.py:137]: Seems to be missing a blank line above the method definition.

File: saas/mixins.py [hunk 4]
```diff
@@ -1215,7 +1216,12 @@ def get_object(self):
 
 class BalancesDueMixin(DateRangeContextMixin, ProviderMixin):
 
-    filter_backends = (DateRangeFilter,) # XXX Why no search and ordering?
+    search_fields = (
+        'slug',
+        'full_name',
+    )
+
+    filter_backends = (DateRangeFilter, SearchFilter, OrderingFilter)
```
Comment (smirolo) [inline saas/mixins.py:1224]: There is an `OrderingFilter` but there does not seem to be a default ordering provided.

File: saas/views/download.py [hunk 5]
```diff
@@ -421,3 +424,273 @@ def queryrow_to_columns(self, record):
             self.encode(transaction.dest_unit),
             self.encode_descr(transaction)
         ]
+
+class BalancesMetricsDownloadView(CSVDownloadView, BalancesMetricsMixin):
```
Comment (smirolo) [inline saas/views/download.py:428]: Mixin should be before View (i.e. View is always the last class in the inheritance)

`class BalancesMetricsDownloadView(BalancesMetricsMixin, CSVDownloadView):`

File: saas/views/download.py [hunk 6]
```diff
@@ -421,3 +424,273 @@ def queryrow_to_columns(self, record):
...
+    ]
+
+    def get_headings(self):
+        return self.headings
+
+    def get_queryset(self):
+        request = Request(self.request)
+        results, unit = self.get_data(request=request)
+
+        consolidated_data = {}
```
Comment (smirolo) [inline saas/views/download.py:445]: What is the purpose of this code block? It looks like a duplicate of some code in the `retrieve_metrics` mixin.

File: saas/views/download.py [hunk 7]
```diff
@@ -421,3 +424,273 @@ def queryrow_to_columns(self, record):
...
+            row = [date] + values
+            csv_data.append(row)
+
+        return csv_data
+
+    def queryrow_to_columns(self, record):
+        return self.encode(record)
+
+    def get(self, request, *args, **kwargs):
+        self.set_plans_data(request)
```
Comment (smirolo) [inline saas/views/download.py:583]: I don't understand the purpose of the complexity here. Is it to define `self.plans`? In that case why not use a property?

```
@property
def plans(self):
    if not hasattr(self, '_plans'):
        self._plans = ...
    return self._plans
```

File: saas/views/download.py [hunk 8]
```diff
@@ -421,3 +424,273 @@ def queryrow_to_columns(self, record):
...
+            for currency in organization.get('balances').keys():
+                currency_set.add(currency)
+        return currency_set
+
+    def get_queryset(self):
+        queryset = super(BalancesDueDownloadView, self).get_queryset()
+        decorated_queryset = super(BalancesDueDownloadView, self).decorate_queryset(
+            queryset)
+        # Using a serializer to get type, credentials
+        serializer = BalancesDueSerializer(decorated_queryset, many=True)
```
Comment (smirolo) [inline saas/views/download.py:625]: Why is the serialization happening in `get_queryset`? That doesn't seem the correct place to do it.

File: saas/views/download.py [hunk 9]
```diff
@@ -421,3 +424,273 @@ def queryrow_to_columns(self, record):
...
+        'Credentials',
+        'Created At'
+    ]
+
+    def get_headings(self):
+        return self.headings
+
+    def get_queryset(self):
+        # Using a serializer to get type & credentials
+        serializer = OrganizationSerializer(super().get_queryset(), many=True)
```
Comment (smirolo) [inline saas/views/download.py:689]: another serialization where it is not expected.

File: saas/api/metrics.py [hunk 10]
```diff
@@ -56,23 +56,178 @@ class MetricsMixin(DateRangeContextMixin, ProviderMixin):
     serializer_class = MetricsSerializer
     filter_backends = (TrailingPeriodFilter,)
 
-    def get(self, request, *args, **kwargs):
+    def calculate_date_periods(self, request):
```
Comment (smirolo) [inline saas/api/metrics.py:59]: Interesting introduction of `calculate_date_periods`. looks good!

**Review by smirolo — COMMENTED** [review]
Need to discuss some of the choices around returned results by `get_queryset`.

**Pull request merged**